maps/src/main/java/org/odk/collect/maps/layers/OfflineMapLayersPicker.kt

```diff
@@ -5,67 +5,109 @@ import android.os.Bundle
             }
         }
     }
 
-    private lateinit var offlineMapLayersPickerBinding: OfflineMapLayersPickerBinding
+    private lateinit var binding: OfflineMapLayersPickerBinding
+
+    private val getLayers = registerForActivityResult(ActivityResultContracts.GetMultipleContents(), registry) { uris ->
+        if (uris.isNotEmpty()) {
+            viewModel.loadLayersToImport(uris)
+            DialogFragmentUtils.showIfNotShowing(
+                OfflineMapLayersImporter::class.java,
+                childFragmentManager
+            )
+        }
+    }
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        childFragmentManager.fragmentFactory = FragmentFactoryBuilder()
+            .forClass(OfflineMapLayersImporter::class) {
+                OfflineMapLayersImporter(referenceLayerRepository, scheduler, settingsProvider)
+            }
+            .build()
+
+        super.onCreate(savedInstanceState)
+    }
 
     override fun onCreateView(
         inflater: LayoutInflater,
         container: ViewGroup?,
         savedInstanceState: Bundle?
     ): View {
-        offlineMapLayersPickerBinding = OfflineMapLayersPickerBinding.inflate(inflater)
-
-        viewModel.data.observe(this) { data ->
-            offlineMapLayersPickerBinding.progressIndicator.visibility = View.GONE
-            offlineMapLayersPickerBinding.layers.visibility = View.VISIBLE
-            offlineMapLayersPickerBinding.save.isEnabled = true
+        binding = OfflineMapLayersPickerBinding.inflate(inflater)
 
-            val offlineMapLayersAdapter = OfflineMapLayersAdapter(data.first, data.second) {
-                viewModel.changeSelectedLayerId(it)
-            }
-            offlineMapLayersPickerBinding.layers.setAdapter(offlineMapLayersAdapter)
-        }
-
-        offlineMapLayersPickerBinding.mbtilesInfoGroup.addOnClickListener {
+        binding.mbtilesInfoGroup.addOnClickListener {
             externalWebPageHelper.openWebPageInCustomTab(
                 requireActivity(),
                 Uri.parse("https://docs.getodk.org/collect-offline-maps/#transferring-offline-tilesets-to-devices")
             )
         }
 
-        offlineMapLayersPickerBinding.cancel.setOnClickListener {
+        binding.addLayer.setOnClickListener {
+            getLayers.launch("*/*")
+        }
+
+        binding.cancel.setOnClickListener {
             dismiss()
         }
 
-        offlineMapLayersPickerBinding.save.setOnClickListener {
+        binding.save.setOnClickListener {
             viewModel.saveSelectedLayer()
             dismiss()
         }
-        return offlineMapLayersPickerBinding.root
+
+        return binding.root
+    }
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+
+        viewModel.isLoading.observe(this) { isLoading ->
```

Comment:
I was going to say that these observations need to use `viewLifecycleOwner`, but I was super confused as to why we weren't getting editor warnings about using `this`. For the 1000th time, I'd forgotten that you use the Fragment itself in `DialogFragment` (https://developer.android.com/guide/fragments/dialogs#lifecycle). 🤦‍♂️.